File: strategies/sma_crossover/python/strategy.py
"""The SMA 9/30 rule, independent from APIs and order execution."""

from collections.abc import Sequence
from dataclasses import dataclass
from decimal import Decimal
from typing import Literal

Signal = Literal["entry", "exit"] | None


@dataclass(frozen=True, slots=True)
class Evaluation:
    fast: Decimal
    slow: Decimal
    signal: Signal


def _mean(values: Sequence[Decimal]) -> Decimal:
    return sum(values, start=Decimal(0)) / Decimal(len(values))


def evaluate(
    closes: Sequence[Decimal],
    fast_window: int = 9,
    slow_window: int = 30,
) -> Evaluation:
    """Return current SMA values and an entry/exit crossover signal.

    At least ``slow_window`` closes calculate the averages. One additional
    close is needed before a crossover can be detected against the previous
    candle.
    """

    if fast_window <= 0 or slow_window <= fast_window:
        raise ValueError("windows must satisfy 0 < fast_window < slow_window")
    if len(closes) < slow_window:
        raise ValueError(f"at least {slow_window} closes are required")

    fast = _mean(closes[-fast_window:])
    slow = _mean(closes[-slow_window:])
    if len(closes) == slow_window:
        return Evaluation(fast, slow, None)

    previous = closes[:-1]
    previous_fast = _mean(previous[-fast_window:])
    previous_slow = _mean(previous[-slow_window:])

    if previous_fast <= previous_slow and fast > slow:
        return Evaluation(fast, slow, "entry")
    if previous_fast >= previous_slow and fast < slow:
        return Evaluation(fast, slow, "exit")
    return Evaluation(fast, slow, None)

File: strategies/sma_crossover/node/strategy.ts
export type Signal = "entry" | "exit" | null;

export type Evaluation = Readonly<{
  fast: number;
  slow: number;
  signal: Signal;
}>;

const mean = (values: readonly number[]): number =>
  values.reduce((total, value) => total + value, 0) / values.length;

/** Pure SMA calculation and crossover rule, independent from the SDK. */
export const evaluate = (
  closes: readonly number[],
  fastWindow = 9,
  slowWindow = 30,
): Evaluation => {
  if (fastWindow <= 0 || slowWindow <= fastWindow) {
    throw new RangeError("windows must satisfy 0 < fastWindow < slowWindow");
  }
  if (closes.length < slowWindow) {
    throw new RangeError(`at least ${slowWindow} closes are required`);
  }

  const fast = mean(closes.slice(-fastWindow));
  const slow = mean(closes.slice(-slowWindow));
  if (closes.length === slowWindow) return { fast, slow, signal: null };

  const previous = closes.slice(0, -1);
  const previousFast = mean(previous.slice(-fastWindow));
  const previousSlow = mean(previous.slice(-slowWindow));
  const signal =
    previousFast <= previousSlow && fast > slow
      ? "entry"
      : previousFast >= previousSlow && fast < slow
        ? "exit"
        : null;

  return { fast, slow, signal };
};
